File: scripts/test_connection.py
#!/usr/bin/env python3
"""
==============================================================================
Database Connection Test Script
==============================================================================
Script: test_connection.py
Purpose: Verify Supabase connection and database setup
Author: Road Explorer Portugal
==============================================================================

This script tests:
1. Environment variables are configured
2. Supabase connection works
3. PostGIS extension is enabled
4. Roads table exists and is accessible
5. Geometry functions work correctly
6. Test data can be queried

Usage:
    python test_connection.py

Prerequisites:
    1. Virtual environment activated
    2. Dependencies installed (uv pip install -r requirements.txt)
    3. .env file configured with credentials
==============================================================================
"""

import os
import sys
from dotenv import load_dotenv
from supabase import create_client, Client

# Load environment variables
load_dotenv()


def print_header(text):
    """Print a formatted header."""
    print("\n" + "=" * 70)
    print(f"  {text}")
    print("=" * 70)


def print_test(test_name, status, message=""):
    """Print test result with formatting."""
    status_symbol = "✅" if status else "❌"
    print(f"{status_symbol} {test_name}")
    if message:
        print(f"   {message}")


def test_environment_variables():
    """Test if environment variables are configured."""
    print_header("1. Testing Environment Variables")

    supabase_url = os.getenv("SUPABASE_URL", "")
    supabase_key = os.getenv("SUPABASE_KEY", "")
    mapbox_token = os.getenv("MAPBOX_TOKEN", "")

    url_valid = bool(supabase_url and supabase_url.startswith("https://"))
    key_valid = bool(supabase_key and len(supabase_key) > 20)
    mapbox_valid = bool(mapbox_token and mapbox_token.startswith("pk."))

    print_test("SUPABASE_URL configured", url_valid,
               f"URL: {supabase_url[:30]}..." if url_valid else "Not found or invalid")
    print_test("SUPABASE_KEY configured", key_valid,
               f"Length: {len(supabase_key)} chars" if key_valid else "Not found or too short")
    print_test("MAPBOX_TOKEN configured", mapbox_valid,
               f"Token: {mapbox_token[:15]}..." if mapbox_valid else "Not found or invalid")

    return url_valid and key_valid


def test_supabase_connection():
    """Test Supabase connection."""
    print_header("2. Testing Supabase Connection")

    try:
        supabase_url = os.getenv("SUPABASE_URL", "")
        supabase_key = os.getenv("SUPABASE_KEY", "")

        supabase = create_client(supabase_url, supabase_key)
        print_test("Supabase client created", True, f"Connected to: {supabase_url}")

        return supabase
    except Exception as e:
        print_test("Supabase client creation", False, f"Error: {e}")
        return None


def test_postgis_extension(supabase: Client):
    """Test if PostGIS extension is enabled."""
    print_header("3. Testing PostGIS Extension")

    try:
        # Query to check if PostGIS is installed
        result = supabase.rpc('postgis_version').execute()

        if result.data:
            print_test("PostGIS extension enabled", True, f"Version: {result.data}")
            return True
        else:
            # Alternative: Try a simple PostGIS function
            test_result = supabase.rpc('st_astext', {
                'geom': 'POINT(0 0)'
            }).execute()

            print_test("PostGIS extension enabled", True,
                      "PostGIS functions are working")
            return True

    except Exception as e:
        print_test("PostGIS extension check", False,
                  f"Note: This is expected if PostGIS functions aren't exposed via RPC")
        # This is not critical - PostGIS may be enabled but not exposed via RPC
        print("   ℹ️  PostGIS may still be working. Continuing tests...")
        return True


def test_roads_table(supabase: Client):
    """Test if roads table exists and is accessible."""
    print_header("4. Testing Roads Table")

    try:
        # Try to count roads
        result = supabase.table('roads').select('id', count='exact').execute()

        count = result.count if hasattr(result, 'count') else len(result.data)

        print_test("Roads table accessible", True, f"Found {count} roads")

        return count > 0
    except Exception as e:
        print_test("Roads table access", False, f"Error: {e}")
        return False


def test_query_road_data(supabase: Client):
    """Test querying road data with all fields."""
    print_header("5. Testing Road Data Query")

    try:
        # Fetch first road with all fields
        result = supabase.table('roads').select('*').limit(1).execute()

        if not result.data:
            print_test("Query road data", False, "No data returned")
            return False

        road = result.data[0]

        # Check essential fields
        required_fields = ['id', 'code', 'name', 'region', 'geometry',
                          'distance_km', 'start_lat', 'start_lon']

        missing_fields = [field for field in required_fields if field not in road]

        if missing_fields:
            print_test("Road data structure", False,
                      f"Missing fields: {', '.join(missing_fields)}")
            return False

        print_test("Query road data", True, f"Sample: {road['code']} - {road['name']}")
        print(f"   Region: {road['region']}")
        print(f"   Distance: {road['distance_km']} km")
        print(f"   Geometry type: {type(road['geometry'])}")

        return True

    except Exception as e:
        print_test("Query road data", False, f"Error: {e}")
        return False


def test_geometry_format(supabase: Client):
    """Test if geometry is returned in correct format."""
    print_header("6. Testing Geometry Format")

    try:
        # Fetch a road with geometry
        result = supabase.table('roads').select('code, geometry').limit(1).execute()

        if not result.data:
            print_test("Geometry format test", False, "No data returned")
            return False

        road = result.data[0]
        geometry = road.get('geometry', '')

        # Check if geometry is WKT LINESTRING format
        is_wkt = isinstance(geometry, str) and geometry.startswith('LINESTRING(')

        if is_wkt:
            # Count points in geometry
            coords = geometry.replace('LINESTRING(', '').replace(')', '')
            point_count = len(coords.split(','))

            print_test("Geometry format (WKT)", True,
                      f"Sample: {geometry[:50]}...")
            print(f"   Points in geometry: {point_count}")
            return True
        else:
            print_test("Geometry format", False,
                      f"Unexpected format: {type(geometry)}")
            print(f"   Value: {geometry}")
            return False

    except Exception as e:
        print_test("Geometry format test", False, f"Error: {e}")
        return False


def test_filter_by_region(supabase: Client):
    """Test filtering roads by region."""
    print_header("7. Testing Region Filtering")

    try:
        # Test filtering by Continental
        result = supabase.table('roads').select('code, region') \
                        .eq('region', 'Continental').execute()

        continental_count = len(result.data)

        print_test("Filter by region", True,
                  f"Found {continental_count} Continental roads")

        if continental_count > 0:
            sample_codes = [r['code'] for r in result.data[:3]]
            print(f"   Sample codes: {', '.join(sample_codes)}")

        return True

    except Exception as e:
        print_test("Filter by region", False, f"Error: {e}")
        return False


def main():
    """Main test execution."""

    print("=" * 70)
    print("  🧪 Road Explorer Portugal - Database Connection Test")
    print("=" * 70)
    print("\nThis script will verify your Supabase setup is working correctly.")

    # Run tests
    tests_passed = 0
    tests_total = 7

    # Test 1: Environment variables
    if not test_environment_variables():
        print("\n❌ Environment variables not configured properly!")
        print("   Please check your .env file and ensure all variables are set.")
        sys.exit(1)
    tests_passed += 1

    # Test 2: Supabase connection
    supabase = test_supabase_connection()
    if not supabase:
        print("\n❌ Cannot connect to Supabase!")
        print("   Please check your SUPABASE_URL and SUPABASE_KEY.")
        sys.exit(1)
    tests_passed += 1

    # Test 3: PostGIS (informational)
    test_postgis_extension(supabase)
    tests_passed += 1

    # Test 4: Roads table
    if not test_roads_table(supabase):
        print("\n❌ Roads table not accessible!")
        print("   Please run schema.sql in Supabase SQL Editor first.")
        sys.exit(1)
    tests_passed += 1

    # Test 5: Query road data
    if not test_query_road_data(supabase):
        print("\n⚠️  Warning: Could not query road data properly.")
        print("   You may need to load test data (test_data.sql).")
    else:
        tests_passed += 1

    # Test 6: Geometry format
    if test_geometry_format(supabase):
        tests_passed += 1

    # Test 7: Filter by region
    if test_filter_by_region(supabase):
        tests_passed += 1

    # Summary
    print_header("Test Summary")
    print(f"\n📊 Tests passed: {tests_passed}/{tests_total}")

    if tests_passed == tests_total:
        print("\n🎉 All tests passed! Your Supabase setup is working correctly.")
        print("\n✅ Next steps:")
        print("   1. Run Python data processing scripts to add roads")
        print("   2. Start frontend development (npm run dev)")
        print("   3. Test frontend connection to Supabase")
    elif tests_passed >= 4:
        print("\n⚠️  Most tests passed. Minor issues detected.")
        print("   Review the failures above and fix if needed.")
    else:
        print("\n❌ Multiple tests failed. Please review your setup.")
        print("   Check the errors above and fix configuration issues.")

    print("=" * 70)

    return 0 if tests_passed >= 4 else 1


if __name__ == "__main__":
    sys.exit(main())

File: frontend/src/utils/testConnection.js
/**
 * Supabase Connection Test Utility
 *
 * This file contains functions to test Supabase connection from the frontend.
 * Can be called from browser console or used in development.
 *
 * Usage:
 *   import { testSupabaseConnection } from './utils/testConnection';
 *   testSupabaseConnection();
 *
 * Or in browser console:
 *   window.testSupabaseConnection()
 */

import { supabase, fetchRoads, wktToGeoJSON } from '../services/api';

/**
 * Test basic Supabase connection and configuration
 */
export const testSupabaseConnection = async () => {
  console.log('\n🧪 Testing Supabase Connection...\n');
  console.log('='.repeat(50));

  let allPassed = true;

  // Test 1: Environment variables
  console.log('\n📋 Test 1: Environment Variables');
  const supabaseUrl = import.meta.env.VITE_SUPABASE_URL;
  const supabaseKey = import.meta.env.VITE_SUPABASE_ANON_KEY;
  const mapboxToken = import.meta.env.VITE_MAPBOX_TOKEN;

  const urlValid = Boolean(supabaseUrl?.startsWith('https://'));
  const keyValid = Boolean(supabaseKey?.length > 20);
  const mapboxValid = Boolean(mapboxToken?.startsWith('pk.'));

  console.log(`${urlValid ? '✅' : '❌'} VITE_SUPABASE_URL: ${urlValid ? supabaseUrl?.substring(0, 30) + '...' : 'Not configured'}`);
  console.log(`${keyValid ? '✅' : '❌'} VITE_SUPABASE_ANON_KEY: ${keyValid ? 'Configured (' + supabaseKey?.length + ' chars)' : 'Not configured'}`);
  console.log(`${mapboxValid ? '✅' : '❌'} VITE_MAPBOX_TOKEN: ${mapboxValid ? mapboxToken?.substring(0, 15) + '...' : 'Not configured'}`);

  if (!urlValid || !keyValid) {
    console.error('\n❌ Environment variables not properly configured!');
    console.log('   Please check your .env file in frontend/ directory');
    return false;
  }

  // Test 2: Supabase client initialization
  console.log('\n📡 Test 2: Supabase Client');
  try {
    console.log('✅ Supabase client initialized');
    console.log(`   Connected to: ${supabaseUrl}`);
  } catch (error) {
    console.error('❌ Supabase client initialization failed:', error);
    allPassed = false;
  }

  // Test 3: Fetch roads from database
  console.log('\n🗄️  Test 3: Fetch Roads from Database');
  try {
    const { data, error } = await fetchRoads();

    if (error) {
      console.error('❌ Error fetching roads:', error);
      console.log('   This may indicate:');
      console.log('   - Database schema not yet applied (run scripts/schema.sql)');
      console.log('   - RLS policies blocking access');
      console.log('   - Network connectivity issues');
      allPassed = false;
    } else if (!data || data.length === 0) {
      console.warn('⚠️  Roads table is empty');
      console.log('   Database connection works, but no data found.');
      console.log('   Run scripts/test_data.sql to add sample roads.');
    } else {
      console.log(`✅ Successfully fetched ${data.length} roads`);
      console.log('   Sample roads:');
      data.slice(0, 3).forEach(road => {
        console.log(`   - ${road.code}: ${road.name} (${road.distance_km}km)`);
      });
    }
  } catch (error) {
    console.error('❌ Fetch roads failed:', error);
    allPassed = false;
  }

  // Test 4: Test geometry conversion
  console.log('\n🗺️  Test 4: Geometry Conversion (WKT to GeoJSON)');
  try {
    const { data } = await supabase
      .from('roads')
      .select('code, geometry')
      .limit(1)
      .single();

    if (data && data.geometry) {
      const geojson = wktToGeoJSON(data.geometry);

      if (geojson && geojson.type === 'Feature') {
        console.log(`✅ WKT to GeoJSON conversion works`);
        console.log(`   Road: ${data.code}`);
        console.log(`   Geometry points: ${geojson.geometry.coordinates.length}`);
        console.log(`   Sample coordinate: [${geojson.geometry.coordinates[0].join(', ')}]`);
      } else {
        console.error('❌ Geometry conversion failed');
        allPassed = false;
      }
    } else {
      console.warn('⚠️  No geometry data to test conversion');
    }
  } catch (error) {
    console.warn('⚠️  Geometry conversion test skipped (no data)');
  }

  // Test 5: Test filtering by region
  console.log('\n🌍 Test 5: Region Filtering');
  try {
    const { data, error } = await supabase
      .from('roads')
      .select('code, region')
      .eq('region', 'Continental')
      .limit(5);

    if (error) {
      console.error('❌ Region filtering failed:', error);
      allPassed = false;
    } else {
      console.log(`✅ Region filtering works`);
      console.log(`   Found ${data?.length || 0} Continental roads`);
      if (data && data.length > 0) {
        console.log(`   Samples: ${data.map(r => r.code).join(', ')}`);
      }
    }
  } catch (error) {
    console.error('❌ Region filtering test failed:', error);
    allPassed = false;
  }

  // Summary
  console.log('\n' + '='.repeat(50));
  if (allPassed) {
    console.log('✅ All tests passed! Supabase connection is working correctly.');
    console.log('\n📝 Next steps:');
    console.log('   1. Add more roads using Python scripts');
    console.log('   2. Start building frontend components');
    console.log('   3. Test map visualization with real data');
  } else {
    console.log('❌ Some tests failed. Please review errors above.');
    console.log('\n🔧 Common fixes:');
    console.log('   1. Run schema.sql in Supabase SQL Editor');
    console.log('   2. Run test_data.sql for sample data');
    console.log('   3. Check .env file has correct credentials');
    console.log('   4. Verify Supabase RLS policies allow public read');
  }
  console.log('='.repeat(50) + '\n');

  return allPassed;
};

/**
 * Quick connection check (minimal output)
 */
export const quickConnectionCheck = async () => {
  try {
    const { data, error } = await supabase.from('roads').select('count', { count: 'exact', head: true });

    if (error) {
      console.error('❌ Connection check failed:', error.message);
      return false;
    }

    console.log('✅ Supabase connection OK');
    return true;
  } catch (error) {
    console.error('❌ Connection check failed:', error);
    return false;
  }
};

// Expose to window for console testing
if (typeof window !== 'undefined') {
  window.testSupabaseConnection = testSupabaseConnection;
  window.quickConnectionCheck = quickConnectionCheck;
}

export default testSupabaseConnection;
